### PR TITLE
MEN-4324: integration test for the file transfer feature

### PR DESCRIPTION
**Pull request by tranchitella**
Changelog: none

Signed-off-by: Fabio Tranchitella <fabio.tranchitella@northern.tech>

### CODE REVIEW

**Comment by tranchitella** [comment]
@mender-test-bot start pipeline

**Comment by tranchitella** [comment]
@mender-test-bot start pipeline

**Comment by tranchitella** [comment]
Failures in the "mender-connect bogus protocol" are expected, everything else is green, including the file transfer tests.

File: tests/tests/test_filetransfer.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+            assert (
+                r.headers.get("Content-Disposition")
+                == 'attachment; filename="random.bin"'
+            )
+            assert r.headers.get("Content-Type") == "application/octet-stream"
+            assert r.headers.get("X-Men-File-Gid") == "0"
+            assert r.headers.get("X-Men-File-Uid") == "0"
+            assert r.headers.get("X-Men-File-Mode") == "600"
+            assert r.headers.get("X-Men-File-Path") == "/tmp/random.bin"
+            assert r.headers.get("X-Men-File-Size") != ""
```
Comment (lluiscampos) [inline tests/tests/test_filetransfer.py:153]: Can't we check the exact size in this case?

File: tests/tests/test_filetransfer.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+import hashlib
+import io
+import os
+import os.path
```
Comment (lluiscampos) [inline tests/tests/test_filetransfer.py:19]: We can remove this import, as we already have `os`

**Review by lluiscampos — APPROVED** [review]
Looks very good. The test logic is well expressed, and I haven't spotted any obvious error.

I recall @merlin-northern being concern about parallel downloads of files; do you guys think we can test this scenario here or are we confident enough about the protocol now?

**Comment by merlin-northern** [comment]
> Looks very good. The test logic is well expressed, and I haven't spotted any obvious error.
> 
> I recall @merlin-northern being concern about parallel downloads of files; do you guys think we can test this scenario here or are we confident enough about the protocol now?

this is something I want to test some more manually, and probably a ticket is coming, so I would for now keep it as is, if it proves to be a bug, then with the ticket resolution we will extend the test, if it makes sense.

File: tests/tests/test_filetransfer.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+            # upload the file
+            r = upload_file(
+                "/tmp/random.bin",
+                open(f.name, "rb"),
+                devid,
+                authtoken,
+                mode="600",
+                uid="0",
+                gid="0",
```
Comment (kacf) [inline tests/tests/test_filetransfer.py:136]: Suggestion: Use some random mode/uid/gid values here, which will test that both the upload and the download of these work.

**Review by kacf — APPROVED** [review]
Very good test!

**Comment by tranchitella** [comment]
@kacf @lluiscampos I addressed your comments here: https://github.com/mendersoftware/integration/pull/1256/commits/9e7d6cfc946903f56c6d72c0d06311e808cec498
@mender-test-bot start pipeline, please

**Comment by tranchitella** [comment]
@mender-test-bot start pipeline

**Comment by tranchitella** [comment]
@mender-test-bot start pipeline

**Comment by tranchitella** [comment]
Manually launched pipeline (because of issues in GitLab, it seems):
https://gitlab.com/Northern.tech/Mender/mender-qa/-/pipelines/270604329

**Comment by tranchitella** [comment]
Failures in `test_demo_artifact` and `test_bogus_proto_message` are spurious.

**Pull request merged**